include/slb/detail/invoke.hpp

```diff
@@ -0,0 +1,384 @@
+/*
+  SLB.Detail.Invoke
+
+  Copyright Michael Park, 2017
+  Copyright Agustin Berge, 2017
+
+  Distributed under the Boost Software License, Version 1.0.
+  (See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
+*/
+
+#ifndef SLB_DETAIL_INVOKE_HPP
+#define SLB_DETAIL_INVOKE_HPP
+
+#include "config.hpp"
+#include "lib.hpp"
+
+#include <functional>
+#include <type_traits>
+#include <utility>
+
+namespace slb {
+namespace detail {
+namespace _invoke {
```

Comment:
I really don't like leading underscores 😞 I'd prefer trailing, if we had to pick one.

Reply:
We need a distinct namespace, and we want `detail::invoke` to remain being `INVOKE`, so perhaps we can rename this one to `invoke_impl`?

Reply:
yeah, I'd prefer `invoke_impl`.